spec/model/postgres/postgres_resource_spec.rb

```diff
@@ -24,121 +38,218 @@
   end
 
   it "returns connection string with ubid qualifier" do
+    postgres_resource.update(hostname_version: "v2")
     expect(postgres_resource).to receive(:dns_zone).and_return("something").at_least(:once)
-    expect(postgres_resource.connection_string).to eq("postgres://postgres:dummy-password@pg-name.pgc60xvcr00a5kbnggj1js4kkq.postgres.ubicloud.com:5432/postgres?channel_binding=require")
+    expect(postgres_resource.connection_string).to eq("postgres://postgres:dummy-password@pg-name.#{postgres_resource.ubid}.postgres.ubicloud.com:5432/postgres?channel_binding=require")
   end
 
   it "returns connection string with ip address if config is not set" do
-    expect(postgres_resource).to receive(:representative_server).and_return(instance_double(PostgresServer, vm: instance_double(Vm, ip4: "1.2.3.4", ip4_string: "1.2.3.4"))).at_least(:once)
+    vm = create_hosted_vm(project, private_subnet, "pg-vm")
+    PostgresServer.create(timeline:, resource_id: postgres_resource.id, vm_id: vm.id, representative_at: Time.now, synchronization_status: "ready", timeline_access: "push", version: "17")
+    AssignedVmAddress.create(dst_vm_id: vm.id, ip: "1.2.3.4/32")
     expect(postgres_resource.connection_string).to eq("postgres://postgres:dummy-password@1.2.3.4:5432/postgres?channel_binding=require")
   end
 
   it "returns connection string as nil if there is no server" do
-    expect(postgres_resource).to receive(:representative_server).and_return(nil).at_least(:once)
+    # No server created, no dns_zone
     expect(postgres_resource.connection_string).to be_nil
   end
 
   it "returns replication_connection_string" do
     s = postgres_resource.replication_connection_string(application_name: "pgubidstandby")
-    expect(s).to include("ubi_replication@pgc60xvcr00a5kbnggj1js4kkq.postgres.ubicloud.com", "application_name=pgubidstandby", "sslcert=/etc/ssl/certs/server.crt")
+    expect(s).to include("ubi_replication@#{postgres_resource.ubid}.postgres.ubicloud.com", "application_name=pgubidstandby", "sslcert=/etc/ssl/certs/server.crt")
   end
 
   it "returns has_enough_fresh_servers correctly" do
-    expect(postgres_resource.servers).to receive(:count).and_return(1, 1)
-    expect(postgres_resource).to receive(:target_server_count).and_return(1, 2)
+    # Stub servers.count block call - actual implementation checks needs_recycling? on each server
+    # which requires complex VM size and storage volume matching
+    expect(postgres_resource).to receive(:servers).and_return(double(count: 1)).twice
```

Comment:
I would push the mocking further down here. `servers` is an association, so it's cached, and you can add expectations on the cached values. I would create a single `PostgresServer` and mock only the `needs_recycling` method.

```suggestion
    # create PostgresServer
    expect(postgres_resources.servers.first).to receive(:needs_recycling?).and_return(false).twice
```

Same in other places where you you are mocking `servers`.

Reply:
this is the ai lazing out, I'll bark at it